# JUPYTER                                     29/7/2024

## Task

Today completed jupyter and Rmarkdown by practising each and every command.

`%lsmagic` - To check magics available

`%line` - for line magic only works for line.

`%%bash` -cell magic works for whole cell.

`from matplotlib import pyplot as plt
import numpy as np

x = np.linspace(1,3*np.pi,100)

y= np.sin(x)

plt.plot(x,y)

plt.show()` - using matplotlib for visualization

`import seaborn as sns`
```
from ipywidgets import interactive 
from matplotlib import pyplot as plt
import numpy as np
def sine_curve(A,f,p):
    plt.figure(1,figsize=(4,4))
    x=np.linspace(1,10,100)
    y=A*np.sin(x*f+p)
    plt.plot(x,y,color="red",linestyle="-")
    plt.show()
```
```
interactive_plot = interactive(sine_curve, 
                               A=(1, 5, 1), 
                               f=(0, 5, 1), 
                               p=(0, 5, 0.5))
interactive_plot
```

creating interactive plot using widgets (from ipywidgets)


`from ipywidgets import widgets`

`jupyter nbconvert --to html Untitled.ipynb`

`jupyter nbconvert --to html --template=nbextensions Biopython.ipynb` 

`jupyter nbconvert --to html --output report.html --execute <nbname>.ipynb` - To execute command thorugh terminal and save that to html.

# Task jupyter 

### By jupyter notebook I have created a VCF file with tab delimited and clear visualization.
```
`import pandas as pd
import gzip`

`vcf_file="/home/bharath/Downloads/variants.vcf.gz"
with gzip.open(vcf_file,'rt') as t:
    df_variants=pd.read_csv(t, sep='\t', comment='#')`

`df_variants.shape`

`vcf_file="/home/bharath/snakemake/mapping/extra/output/file.sorted.dedup.recal.bqsr.haplotype.vcf"
df_variants=pd.read_csv(vcf_file,sep='\t',comment='#')`

`import os
cmd="zgrep '^#' "+ vcf_file +" | tail -n 1"`

`columns=os.popen(cmd).read()`

`columns=columns.strip('#')`

`columns=columns.strip("\n")`

`columns=columns.split("\t")`

`df_variants=pd.read_csv(vcf_file,sep='\t',comment='#',names=columns)`
```

## Snakemake

`snakemake -n -r -p <final_output>` - For dry run.

`snakemake --cores 1` - to run the snakefile.

`snakemake --rulegraph output/rna_featurecounts.txt | dot -Tpng > rulegraph.png` - To create a DAG(directed acyclic graph) by using this command.

`snakemake --dag output/rna_featurecounts.txt | dot -Tpng > jobgraph.png` - To create job graph

`snakemake -n -r output/rna_featurecounts.txt` - To check new outputs included or not.

`snakemake --cores 1 -D > summary.tsv` - To get the summary of the rules used.

`snakemake -s <name>.smk | dot Tpng > <image>.png` - To get overall architecture in single command.

`params` - Helps to pass parameters without changing the script.

`snakemake -R .smk` - -R indicates rerun the smk file.

`config` - To pass the parameter through config file.

`snakemake --config config.yml` - command to add config file to snakemake.

`log` - For debugging.

`snakemake -F` - For reexecution.

`snakemake -D` - For add summary.

`snakemake -S` - For short summary.

`temp` - by marking the output as temp it will delete after the usage of the data.

`protected` - It is away from accidental deletion.

`ancient` - It consider as old output generated after several process.

`touch` 

`directory` - The output file is a directory.

`snakemake -R` - To rerun the snakefile.

`expand` - This function used to specify the multiple input files from the user in more precise way.

`shadow: minimal` - To create isolate environment.

## NEXTFLOW

`nextflow run main.nf` - To run the nextflow script.

`.nextflow.log` - To check for debugging.

`-resume` - To rerun from the discontinued task.

`nextflow run main.nf -with-report -with-timeline -with-dag dag.png` - To get the output with report, time for each process, graphical representation.

`nextflow.enable.dsl=2` - enables more functionality.

`tag ${sra_id}` - Labels the process with current SRA id.

`publishDir, mode: 'copy'` - copies the output to the specified directory.

`val (sra_id)` - expected SRA id.

`nextflow run main.nf -with-docker <image>` - Running docker with container.



