# JUPYTER                                     29/7/2024

## Task

Today completed jupyter and Rmarkdown by practising each and every command.

`%lsmagic` - To check magics available

`%line` - for line magic only works for line.

`%%bash` -cell magic works for whole cell.

`from matplotlib import pyplot as plt
import numpy as np

x = np.linspace(1,3*np.pi,100)

y= np.sin(x)

plt.plot(x,y)

plt.show()` - using matplotlib for visualization

`import seaborn as sns`
```
from ipywidgets import interactive 
from matplotlib import pyplot as plt
import numpy as np
def sine_curve(A,f,p):
    plt.figure(1,figsize=(4,4))
    x=np.linspace(1,10,100)
    y=A*np.sin(x*f+p)
    plt.plot(x,y,color="red",linestyle="-")
    plt.show()
```
```
interactive_plot = interactive(sine_curve, 
                               A=(1, 5, 1), 
                               f=(0, 5, 1), 
                               p=(0, 5, 0.5))
interactive_plot
```

creating interactive plot using widgets (from ipywidgets)


`from ipywidgets import widgets`

`jupyter nbconvert --to html Untitled.ipynb`

`jupyter nbconvert --to html --template=nbextensions Biopython.ipynb` 

`jupyter nbconvert --to html --output report.html --execute <nbname>.ipynb` - To execute command thorugh terminal and save that to html.

# Task jupyter 

### By jupyter notebook I have created a VCF file with tab delimited and clear visualization.
```
`import pandas as pd
import gzip`

`vcf_file="/home/bharath/Downloads/variants.vcf.gz"
with gzip.open(vcf_file,'rt') as t:
    df_variants=pd.read_csv(t, sep='\t', comment='#')`

`df_variants.shape`

`vcf_file="/home/bharath/snakemake/mapping/extra/output/file.sorted.dedup.recal.bqsr.haplotype.vcf"
df_variants=pd.read_csv(vcf_file,sep='\t',comment='#')`

`import os
cmd="zgrep '^#' "+ vcf_file +" | tail -n 1"`

`columns=os.popen(cmd).read()`

`columns=columns.strip('#')`

`columns=columns.strip("\n")`

`columns=columns.split("\t")`

`df_variants=pd.read_csv(vcf_file,sep='\t',comment='#',names=columns)`
```

## Snakemake

`snakemake -n -r -p <final_output>` - For dry run.

`snakemake --cores 1` - to run the snakefile.

`snakemake --rulegraph output/rna_featurecounts.txt | dot -Tpng > rulegraph.png` - To create a DAG(directed acyclic graph) by using this command.

`snakemake --dag output/rna_featurecounts.txt | dot -Tpng > jobgraph.png` - To create job graph

`snakemake -n -r output/rna_featurecounts.txt` - To check new outputs included or not.

`snakemake --cores 1 -D > summary.tsv` - To get the summary of the rules used.

`snakemake -s <name>.smk | dot Tpng > <image>.png` - To get overall architecture in single command.

`params` - Helps to pass parameters without changing the script.

`snakemake -R .smk` - -R indicates rerun the smk file.

`config` - To pass the parameter through config file.

`snakemake --config config.yml` - command to add config file to snakemake.

`log` - For debugging.

`snakemake -F` - For reexecution.

`snakemake -D` - For add summary.

`snakemake -S` - For short summary.

`temp` - by marking the output as temp it will delete after the usage of the data.

`protected` - It is away from accidental deletion.

`ancient` - It consider as old output generated after several process.

`touch` 

`directory` - The output file is a directory.

`snakemake -R` - To rerun the snakefile.

`expand` - This function used to specify the multiple input files from the user in more precise way.

`shadow: minimal` - To create isolate environment.

```
rule all:
    input:
        "rna_featurecounts.txt"

rule QC:
    input:
        fastq="/home/bharath/pipelines/rna_pipeline/fastq/rna.fastq"        
    output:
        "rna_fastqc.zip",
        "rna_fastqc.html"
    shell:
        "fastqc {input.fastq} -o output/"

rule trim:
    input:
        trimjar="/home/bharath/pipelines/Trimmomatic-0.39/trimmomatic-0.39.jar",
        fastq="/home/bharath/pipelines/rna_pipeline/fastq/rna.fastq"   
    
    output:
        "rna_trimmed.fastq"
    
    
    shell:
        "java -jar {input.trimjar} SE -threads 4 {input.fastq} rna_trimmed.fastq TRAILING:10 -phred33"
        
rule hisat:
    input:
        index="/home/bharath/pipelines/rna_pipeline/grch38",
        fastq="rna_trimmed.fastq"
    
    output:
        "rna_trimmed.bam"
    
    shell:
        "hisat2 -q --rna-strandness R -x {input.index}/genome -U {input.fastq} | samtools sort -o rna_trimmed.bam"
        
rule featurecounts:
    input:
        annotation="/home/bharath/pipelines/rna_pipeline/Homo_sapiens.GRCh38.106.gtf",
        bam="rna_trimmed.bam"
    
    output:
        "rna_featurecounts.txt"
    
    shell:
        "featureCounts -S 2 -a {input.annotation} -o rna_featurecounts.txt {input.bam}"

```
**Result**

![RNA ANALYSIS SNAKEMAKE](https://github.com/user-attachments/assets/be8c6faa-2aec-43ca-b715-ce262bd465ba)

## NEXTFLOW

`nextflow run main.nf` - To run the nextflow script.

`.nextflow.log` - To check for debugging.

`-resume` - To rerun from the discontinued task.

`nextflow run main.nf -with-report -with-timeline -with-dag dag.png` - To get the output with report, time for each process, graphical representation.

`nextflow.enable.dsl=2` - enables more functionality.

`tag ${sra_id}` - Labels the process with current SRA id.

`publishDir, mode: 'copy'` - copies the output to the specified directory.

`val (sra_id)` - expected SRA id.

`nextflow run main.nf -with-docker <image>` - Running docker with container.

# TASK NEXTFLOW

```
params.fastq="/home/bharath/pipelines/rna_pipeline/fastq/rna.fastq"
params.index="/home/bharath/pipelines/rna_pipeline/grch38/genome"
params.annotation="/home/bharath/pipelines/rna_pipeline/Homo_sapiens.GRCh38.106.gtf"
params.trimjar="/home/bharath/pipelines/Trimmomatic-0.39/trimmomatic-0.39.jar"
params.output="/home/bharath/pipelines/rna_pipeline/output"

workflow {

    process QC {
        publishDir("${params.output}", mode: 'copy')
        
        input:
        path fastq
        
        output:
        path "*"
        
        script:
        """
        fastqc ${fastq} 
        """
    }

    process trimmomatic {
        publishDir("${params.output}", mode: 'copy')
        
        input:
        val jar
        path fastq
        
        output:
        path "${fastq.baseName}.demo_trimmed.fastq"
        
        script:
        """
        java -jar ${params.trimjar} SE -threads 4 ${fastq} ${fastq.baseName}.demo_trimmed.fastq TRAILING:10 -phred33
        """
    }

    process hisat {
        publishDir("${params.output}", mode: 'copy')
        
        input:
        path index
        path fastq
        
        output:
        path "${fastq.baseName}.demo_trimmed.bam"
        
        script:
        """
        hisat2 -q --rna-strandness R -x ${params.index} -U ${params.fastq} | samtools sort -o ${fastq.baseName}.demo_trimmed.bam
        """
    }

    process featurecount {
        publishDir("${params.output}", mode: 'copy')
        
        input:
        path annotation
        path bam
        
        output:
        path "rna_featurecounts.txt"
        
        script:
        """
        featureCounts -S 2 -a ${params.annotation} -o rna_featurecounts.txt ${bam}
        """
    }

    index_ch = Channel.of(params.index)
    annotation_ch = Channel.of(params.annotation)
    fastq_ch = Channel.of(params.fastq)
    trim_ch = Channel.of(params.trimjar)
    quality_control = QC(fastq_ch)
    trimadapters = trimmomatic(trim_ch, fastq_ch)
    alignment = hisat(trimadapters, index_ch)
    featurecount(annotation_ch, alignment)

}

```
![RNA ANALYSIS NEXTFLOW](https://github.com/user-attachments/assets/f162a9d0-f3c7-4f6a-8a7e-3ac38abb000b)



